### PR TITLE
Add INT8 support for fused_multi_transformer_op

### PR DESCRIPTION
**Pull request by RichardWooSJTU**
### PR types

New features
### PR changes

OPs
### Describe

Add fused_multi_transformer_int8 op to support quantization inference without trt. The reason of quantization inference using native inference instead of trt is two tensorRt engines introduced by while op cannot share weights which cause double GPU memory.  With native inference, we can manage weights flexibly,  but the inference performance is slightly inferior compared with trt.  To gain better performance, we made the following attempts:
1.  Use cublasLt int8 gemm with imma kernel as backend of fc layer.
2. Modify AttnLayernorm/FusedDropoutHelper/FusedDropoutLayerNormHelper to coexist with quantization/de-quantization to speed up inference. We reuse the original classes and functions while adding 2 typenames (InType, OutType) to indicate whether using quantization or dequantization, 3 arguements (quant_out_scale_data, quant_out_scale_offset, quant_in_scale_data) to help Q/DQ. Some notes:
a. We define the above 5 arguements with default values, which means we didn't need to modify the existed references.
b. The above changes only apply to dropout-rate != 1.0 and the 3 classes. 
c. Only pre-layernorm is fully tested.

Some limitations by now:
1. The Inference model generated by paddle.static.save_inference_model() directly (which is not be fused and whose weight is not be quantized) cannot be used by this function. We are working on it and it will be supported soon.
2. Two optimizations can be applied further to enhance the performance (by might only a small margin): 
    a. Batched GEMM are not quantized. 
    b. quant/dequant is explictly called in the QKV GEMM. Fusion into prev/after might be useful.

### CODE REVIEW

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+    tmp.z =
+        __float2int_rn(static_cast<float>(input[m_id * n + n_id + 2]) * scale);
+    tmp.w =
+        __float2int_rn(static_cast<float>(input[m_id * n + n_id + 3]) * scale);
+    output[(m_id * n + n_id) >> 2] = tmp;
+  }
+}
+
+template <typename T>
+void quantize_kernelLauncher(const T* input,
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: quantize_kernel_launcher

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 2]
```diff
@@ -0,0 +1,324 @@
...
+    char4 tmp;
+    tmp.x = __float2int_rn(static_cast<float>(input[m_id * n + n_id]) * scale);
+    tmp.y =
+        __float2int_rn(static_cast<float>(input[m_id * n + n_id + 1]) * scale);
+    tmp.z =
+        __float2int_rn(static_cast<float>(input[m_id * n + n_id + 2]) * scale);
+    tmp.w =
+        __float2int_rn(static_cast<float>(input[m_id * n + n_id + 3]) * scale);
+    output[(m_id * n + n_id) >> 2] = tmp;
+  }
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: m，n较大时一个线程只计算4个可能效率会低，最好这里考虑通用性

File: paddle/fluid/operators/fused/attn_gemm_int8.h (hunk 1 above)
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: 命名不规范

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+                             const float scale,
+                             const int m,
+                             const int n,
+                             cudaStream_t stream) {
+  // TODO(minghaoBD): optimize the kennel launch times when m==1 or n==1
+  dim3 grid((n + 31) / 32, (m + 31) / 32);
+  dim3 block(32, 32);
+
+  quantize_kernel<<<grid, block, 0, stream>>>(
+      input, (char4*)output, scale, m, n);  // NOLINT
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: 同上m、n较大时这里不高效

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+                                  const float* quant_out_scale_data,
+                                  const int layer_offset) {
+  int m_id = blockIdx.x * blockDim.x + threadIdx.x;  // hidden
+  int n_id = blockIdx.y * blockDim.y + threadIdx.y;  // batch size
+
+  bool check = ((m_id < m) && (n_id < n));
+  if (check) {
+    float out_scale = quant_out_scale_data[layer_offset + m_id];
+    output[n_id * m + m_id] =
+        static_cast<T>(static_cast<float>(input[n_id * m + m_id]) * out_scale);
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: 同上

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+}
+
+// dequantize using weight scales and input scales
+template <typename T>
+__global__ void dequantize_kernel(T* output,
+                                  const int32_t* input,
+                                  const int m,  // hidden
+                                  const int n,  // batch size
+                                  const float* quant_out_scale_data,
+                                  const int layer_offset) {
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: layer_offset命名不直观

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+                               const int layer_offset) {
+  dim3 grid((hidden_units + 31) / 32, (batch_size + 31) / 32);
+  dim3 block(32, 32);
+
+  dequantize_kernel<<<grid, block, 0, stream>>>(output,
+                                                input,
+                                                hidden_units,
+                                                batch_size,
+                                                quant_out_scale_data,
+                                                layer_offset);
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: 问题同上

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+template <typename T>
+class AttnMatmulINT8 {
+ public:
+  AttnMatmulINT8(
+      const phi::GPUContext& dev_ctx, int m, int n, int k, bool compute_bias)
+      : dev_ctx_(dev_ctx), m_(m), n_(n), k_(k), compute_bias_(compute_bias) {
+    auto helper = std::make_shared<CublasLtHelper>(m, k, n);
+    helpers_.emplace_back(helper);
+  }
+  ~AttnMatmulINT8() {}
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h:40]: 这里命名INT8的话，上面Q命名也改成INT8

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+
+  void ComputeForward(
+      const framework::Tensor*
+          weight,  // [int8] which has been transformed in pass
```
Comment (qingqing01) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: 这里格式有点乱

review了部分，后续清理后在review~

File: paddle/fluid/operators/fused/attn_gemm_int8.h (hunk 1 above)
Comment (minghaoBD) [reply paddle/fluid/operators/fused/attn_gemm_int8.h]: DONE

Comment (minghaoBD) [reply paddle/fluid/operators/fused/attn_gemm_int8.h]: DONE

File: paddle/fluid/operators/fused/attn_gemm_int8.h (hunk 5 above)
Comment (minghaoBD) [reply paddle/fluid/operators/fused/attn_gemm_int8.h]: rename to: quant_out_scale_offset

File: paddle/fluid/operators/fused/attn_gemm_int8.h (hunk 6 above)
Comment (minghaoBD) [reply paddle/fluid/operators/fused/attn_gemm_int8.h]: DONE

File: paddle/fluid/operators/fused/attn_gemm_int8.h (hunk 7 above)
Comment (minghaoBD) [reply paddle/fluid/operators/fused/attn_gemm_int8.h:40]: DONE. 

As for the fused layernorm-quantization kernel, I am still trying to git rid of the redundant code.

File: paddle/fluid/operators/fused/attn_gemm_int8.h (hunk 8 above)
Comment (minghaoBD) [reply paddle/fluid/operators/fused/attn_gemm_int8.h]: Removed some useless comments.

File: paddle/fluid/operators/fused/quant_dequant_kernel.h [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+#include <vector>
+#include "paddle/fluid/platform/device/gpu/gpu_info.h"
+#include "paddle/fluid/platform/float16.h"
+
+namespace paddle {
+namespace operators {
+
+template <typename T>
+__forceinline__ __device__ int8_t clip_round(const T input, const float scale) {
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/fused/quant_dequant_kernel.h]: 为什么不直接叫"quant"呢？clip_round并不能完整代表该方法的功能

File: paddle/fluid/operators/fused/quant_dequant_kernel.h [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+  float quant_value = 127.0f * (1.0f / scale) * static_cast<float>(input);
+  quant_value = static_cast<float>(round(quant_value));
+  quant_value = quant_value > 127.0f ? 127.0f : quant_value;
+  quant_value = quant_value < -127.0f ? -127.0f : quant_value;
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/fused/quant_dequant_kernel.h]: 这里确认下是否应该改成`-128`

参考这里：https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/fluid/operators/fake_quantize_op.cu.h#L240-L243

File: paddle/fluid/operators/fused/quant_dequant_kernel.h [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+// dequantize using weight scales and input scales
+template <typename T>
+__global__ void dequantize_kernel(T* output,
+                                  const int32_t* input,
+                                  const int m,  // hidden
+                                  const int n,  // batch size
+                                  const float quant_in_scale,
+                                  const float* quant_out_scale_data,
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/fused/quant_dequant_kernel.h]: dequant理论上乘一个dequant_scale就行了，其中，dequant_scale = intput_scale * weight_scale
quant_in_scale和quant_out_scale各是什么意思？

File: paddle/fluid/operators/fused/quant_dequant_kernel.h [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+                                  const int quant_out_scale_offset) {
+  int m_id = blockIdx.x * blockDim.x + threadIdx.x;  // hidden
+  int n_id = blockIdx.y * blockDim.y + threadIdx.y;  // batch size
+
+  bool check = ((m_id < m) && (n_id < n));
+  if (check) {
+    float out_scale = quant_out_scale_data[quant_out_scale_offset + m_id];
+    output[n_id * m + m_id] =
+        static_cast<T>(static_cast<float>(input[n_id * m + m_id]) *
+                       quant_in_scale / out_scale);
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/fused/quant_dequant_kernel.h:110]: 看起来是dequant+quant?

File: paddle/fluid/operators/fused/quant_dequant_kernel.h [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+}
+
+template <typename T>
+void dequantize_kernel_launcher(const int32_t* input,
+                                T* output,
+                                const int batch_size,    // m
+                                const int hidden_units,  // n
+                                cudaStream_t stream,
+                                const float quant_in_scale,
+                                const float* quant_out_scale_data,
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/fused/quant_dequant_kernel.h]: 为什么quant_out_scale_data会有多个数呢？channel-wise dequant?

File: paddle/fluid/operators/fused/attention_layer_norm.h [hunk 14]
```diff
@@ -33,25 +34,36 @@ class AttnLayerNorm {
...
                       const LayerNormParamType<T>* scale_data,
                       const LayerNormParamType<T>* bias_data,
-                      T* y_data,
+                      OutType* y_data,
                       LayerNormParamType<T>* mean_data,
-                      LayerNormParamType<T>* var_data) {
+                      LayerNormParamType<T>* var_data,
+                      const float* quant_out_scale_data = nullptr,
+                      const int quant_out_scale_offset = 0,
+                      const float quant_in_scale_data = 1.0) {
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/fused/attention_layer_norm.h]: 如果不是指针，直接命名为`quant_in_scale`?

File: paddle/fluid/operators/fused/fused_multi_transformer_int8_op.cu [hunk 15]
```diff
@@ -0,0 +1,634 @@
...
+
+    // input scales, vector, size = num_layers
+    auto qkv_in_scale = ctx.Attr<std::vector<float>>("qkv_in_scale");
+    auto out_linear_in_scale =
+        ctx.Attr<std::vector<float>>("out_linear_in_scale");
+    auto ffn1_in_scale = ctx.Attr<std::vector<float>>("ffn1_in_scale");
+    auto ffn2_in_scale = ctx.Attr<std::vector<float>>("ffn2_in_scale");
+
+    // output scales, tensor, size = [num_layers, n], n is gemm output size
+    auto *qkv_out_scale = ctx.Input<Tensor>("QKVOutScale");
```
Comment (wanghaoshuang) [inline paddle/fluid/operators/fused/fused_multi_transformer_int8_op.cu:51]: 这个scale是`1/weights_scale`么？
1. 在量化旧格式中，将weight scale存在了out scale中，这么做不太合理
2. 不应该将格式的约束带入推理实现中，换句话说，推理的实现应该独立于量化模型格式。
3. 应该通过pass来解耦量化格式和推理实现。在Pass中，拿到dequant需要的所有信息，并计算出dequant scales。只需要向推理Deqaunt Operator传递dequant scales。

File: paddle/fluid/operators/fused/attn_gemm_int8.h [hunk 16]
```diff
@@ -0,0 +1,176 @@
...
+                      const framework::Tensor* input,
+                      framework::Tensor* input_tmp,
+                      const framework::Tensor* bias,
+                      framework::Tensor* output,
+                      framework::Tensor* output_tmp,
+                      framework::Tensor* bias_out,
+                      const float quant_in_scale_data,
+                      const framework::Tensor* quant_out_scale,
+                      const int quant_out_scale_offset) {
+    int m = m_, k = k_, n = n_;
```
Comment (heavengate) [inline paddle/fluid/operators/fused/attn_gemm_int8.h]: 这里的m n k好像没有被用到

File: paddle/fluid/operators/fused/cublaslt.h [hunk 17]
```diff
@@ -0,0 +1,182 @@
...
+    if (A_desc_) dyl::cublasLtMatrixLayoutDestroy(A_desc_);
+    if (B_desc_) dyl::cublasLtMatrixLayoutDestroy(B_desc_);
+    if (C_desc_) dyl::cublasLtMatrixLayoutDestroy(C_desc_);
+  }
+
+  void GEMM(int8_t* A_dev,
+            const int8_t* B_dev,
+            int32_t* C_dev,
+            cudaStream_t stream) {
+    // PADDLE_ENFORCE_GPU_SUCCESS(cudaDeviceSynchronize());
```
Comment (heavengate) [inline paddle/fluid/operators/fused/cublaslt.h]: 确认下是否需要，不需要的话可以删除

File: paddle/fluid/operators/fused/quant_dequant_kernel.h (hunk 9 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/quant_dequant_kernel.h]: 已修改为quant_helper

File: paddle/fluid/operators/fused/quant_dequant_kernel.h (hunk 10 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/quant_dequant_kernel.h]: 这里与fake_quant_abs_max_op对齐，确认是-127 
参考：https://github.com/PaddlePaddle/Paddle/blob/b042a3b1736b04cc95faee6d8df118051bef8dad/paddle/fluid/operators/fake_quantize_op.cu.h#L245


同时为了部分解耦，clip不直接hard code 127.0f而是使用max_bound/min_bound参数，该参数为op的属性

File: paddle/fluid/operators/fused/quant_dequant_kernel.h (hunk 11 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/quant_dequant_kernel.h]: quant_in_scale 为op 属性中的input scale，与PTQ导出的scale意义相同
quant_out_scale现改名为dequant_out_scale，为op输入中的output scale，定义与fake_dequant_range_abs_max op中的max_range属性意义相同。
op的属性、输入的定义在fused_multi_transformer_int8_op.cc中进行了文字说明。

File: paddle/fluid/operators/fused/quant_dequant_kernel.h (hunk 12 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/quant_dequant_kernel.h:110]: 为了与fake_dequant_abs_max op对齐以提高精度的暂时的操作，参考：
https://github.com/PaddlePaddle/Paddle/blob/92e1f64b7dcee28e93cfdd984ead0e42e552b491/paddle/fluid/operators/fake_dequantize_op.cu.h#L29

File: paddle/fluid/operators/fused/quant_dequant_kernel.h (hunk 13 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/quant_dequant_kernel.h]: 是channel-wise dequant，也可以兼容layer-wise dequant，每个channel值相同即可

File: paddle/fluid/operators/fused/attention_layer_norm.h (hunk 14 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/attention_layer_norm.h]: 已修改

File: paddle/fluid/operators/fused/fused_multi_transformer_int8_op.cu (hunk 15 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/fused_multi_transformer_int8_op.cu:51]: 已沟通

File: paddle/fluid/operators/fused/attn_gemm_int8.h (hunk 16 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/attn_gemm_int8.h]: 已修改

File: paddle/fluid/operators/fused/cublaslt.h (hunk 17 above)
Comment (RichardWooSJTU) [reply paddle/fluid/operators/fused/cublaslt.h]: 已删除

**Review by Aurelius84 — APPROVED** [review]
LGTM for data registeration

**Review by XieYunshen — APPROVED** [review]
LGTM
单测时间设置

**Review by qingqing01 — APPROVED** [review]

**Review by YuanRisheng — APPROVED** [review]

**Review by XiaoguangHu01 — APPROVED** [review]
LGTM

**Pull request merged**